File: samples/signature-verification/webhook_verifier.py
"""
WebhookEngine — Signature Verification Helper (Python)

Copy this file into your project to verify webhook signatures.
No external dependencies required — uses only the Python standard library.

Usage (Flask):

    from webhook_verifier import verify_webhook

    @app.post("/webhook")
    def handle_webhook():
        is_valid = verify_webhook(
            webhook_id=request.headers.get("webhook-id", ""),
            webhook_timestamp=request.headers.get("webhook-timestamp", ""),
            webhook_signature=request.headers.get("webhook-signature", ""),
            body=request.get_data(as_text=True),
            secret=os.environ["WEBHOOK_SECRET"],
        )

        if not is_valid:
            return {"error": "Invalid signature"}, 401
        # Process webhook...
        return {"status": "received"}

Usage (FastAPI):

    from webhook_verifier import verify_webhook
    from fastapi import Request, HTTPException

    @app.post("/webhook")
    async def handle_webhook(request: Request):
        body = (await request.body()).decode("utf-8")
        is_valid = verify_webhook(
            webhook_id=request.headers.get("webhook-id", ""),
            webhook_timestamp=request.headers.get("webhook-timestamp", ""),
            webhook_signature=request.headers.get("webhook-signature", ""),
            body=body,
            secret=os.environ["WEBHOOK_SECRET"],
        )

        if not is_valid:
            raise HTTPException(status_code=401, detail="Invalid signature")
        return {"status": "received"}
"""

import base64
import hashlib
import hmac
import time


def verify_webhook(
    webhook_id: str,
    webhook_timestamp: str,
    webhook_signature: str,
    body: str,
    secret: str,
    tolerance_seconds: int = 300,
) -> bool:
    """
    Verify a WebhookEngine webhook signature.

    Args:
        webhook_id:          Value of the 'webhook-id' header.
        webhook_timestamp:   Value of the 'webhook-timestamp' header (Unix seconds).
        webhook_signature:   Value of the 'webhook-signature' header (e.g. "v1,base64...").
        body:                The raw request body as a string.
        secret:              The signing secret for the application/endpoint.
        tolerance_seconds:   Timestamp tolerance in seconds. Defaults to 300 (5 minutes).

    Returns:
        True if the signature is valid and the timestamp is within tolerance.
    """
    if not all([webhook_id, webhook_timestamp, webhook_signature, secret]):
        return False

    # Check timestamp tolerance
    try:
        ts = int(webhook_timestamp)
    except (ValueError, TypeError):
        return False

    now = int(time.time())
    drift = abs(now - ts)
    if drift > tolerance_seconds:
        return False

    # Compute expected signature
    signed_content = f"{webhook_id}.{webhook_timestamp}.{body}"
    secret_bytes = _resolve_secret_bytes(secret)
    digest = hmac.new(secret_bytes, signed_content.encode("utf-8"), hashlib.sha256).digest()
    expected_signature = f"v1,{base64.b64encode(digest).decode('utf-8')}"

    # The header may contain multiple signatures separated by spaces
    signatures = webhook_signature.split(" ")
    for sig in signatures:
        # Use constant-time comparison to prevent timing attacks
        if hmac.compare_digest(sig.strip(), expected_signature):
            return True

    return False


def _resolve_secret_bytes(secret: str) -> bytes:
    """
    Resolve the signing secret to bytes.
    Supports whsec_ prefix (Standard Webhooks format), base64-encoded, and raw strings.
    """
    # Support whsec_ prefix
    if secret.lower().startswith("whsec_"):
        return secret.encode("utf-8")

    # Try base64 decoding first
    try:
        decoded = base64.b64decode(secret, validate=True)
        # Verify it's actually valid base64 (round-trip check)
        if base64.b64encode(decoded).decode("utf-8") == secret:
            return decoded
    except Exception:
        pass

    return secret.encode("utf-8")

File: samples/signature-verification/webhook_verifier.ts
// ---------------------------------------------------------------
// WebhookEngine — Signature Verification Helper (TypeScript)
// ---------------------------------------------------------------
// Copy this file into your project to verify webhook signatures.
// Works with Node.js (built-in crypto module) — no npm packages needed.
//
// Usage (Express):
//   import { verifyWebhook } from './webhook_verifier';
//
//   app.post('/webhook', express.raw({ type: '*/*' }), (req, res) => {
//     const isValid = verifyWebhook({
//       webhookId:        req.headers['webhook-id'] as string,
//       webhookTimestamp:  req.headers['webhook-timestamp'] as string,
//       webhookSignature:  req.headers['webhook-signature'] as string,
//       body:              req.body.toString('utf-8'),
//       secret:            process.env.WEBHOOK_SECRET!,
//     });
//
//     if (!isValid) return res.status(401).json({ error: 'Invalid signature' });
//     // Process webhook...
//     res.json({ status: 'received' });
//   });
// ---------------------------------------------------------------

import { createHmac, timingSafeEqual } from "crypto";

interface VerifyOptions {
  /** Value of the 'webhook-id' header */
  webhookId: string;
  /** Value of the 'webhook-timestamp' header (Unix seconds) */
  webhookTimestamp: string;
  /** Value of the 'webhook-signature' header (e.g. "v1,base64...") */
  webhookSignature: string;
  /** The raw request body as a string */
  body: string;
  /** The signing secret for the application/endpoint */
  secret: string;
  /** Timestamp tolerance in seconds. Defaults to 300 (5 minutes). */
  toleranceSeconds?: number;
}

/**
 * Verifies a WebhookEngine webhook signature.
 * Returns true if the signature is valid and the timestamp is within tolerance.
 */
export function verifyWebhook(options: VerifyOptions): boolean {
  const {
    webhookId,
    webhookTimestamp,
    webhookSignature,
    body,
    secret,
    toleranceSeconds = 300,
  } = options;

  if (!webhookId || !webhookTimestamp || !webhookSignature || !secret) {
    return false;
  }

  // Check timestamp tolerance
  const ts = parseInt(webhookTimestamp, 10);
  if (isNaN(ts)) return false;

  const now = Math.floor(Date.now() / 1000);
  const drift = Math.abs(now - ts);
  if (drift > toleranceSeconds) return false;

  // Compute expected signature
  const signedContent = `${webhookId}.${webhookTimestamp}.${body}`;
  const secretBytes = resolveSecretBytes(secret);
  const hash = createHmac("sha256", secretBytes)
    .update(signedContent, "utf-8")
    .digest("base64");
  const expectedSignature = `v1,${hash}`;

  // The header may contain multiple signatures separated by spaces
  const signatures = webhookSignature.split(" ");
  const expectedBuf = Buffer.from(expectedSignature, "utf-8");

  for (const sig of signatures) {
    const sigBuf = Buffer.from(sig.trim(), "utf-8");

    // Use constant-time comparison to prevent timing attacks
    if (
      sigBuf.length === expectedBuf.length &&
      timingSafeEqual(sigBuf, expectedBuf)
    ) {
      return true;
    }
  }

  return false;
}

function resolveSecretBytes(secret: string): Buffer {
  // Support whsec_ prefix (Standard Webhooks format)
  if (secret.toLowerCase().startsWith("whsec_")) {
    return Buffer.from(secret, "utf-8");
  }

  // Try base64 decoding first
  const base64Decoded = Buffer.from(secret, "base64");
  if (base64Decoded.toString("base64") === secret) {
    return base64Decoded;
  }

  return Buffer.from(secret, "utf-8");
}
